Switching cluster causes InterruptedIOException in log

steps:
1. ASSERT: have CRC running
2. EXEC: create context for developer & admin by logging in with both
```
oc login -u developer -p developer https://api.crc.testing:6443
oc login -u kubeadmin -p XXXXXXXXXX https://api.crc.testing:6443
```
3. EXEC: launch plugin
4. ASSERT: context1 active (ex. developer)
5. EXEC: expand Workload > Pods
6. EXEC: switch to context2 (ex. admin)
7. EXEC: open IntelliJ log via Help > Show Log in Finder/Explorer & open log in Editor

Result:
```
2020-10-12 16:23:59,862 [ 112272]   WARN - nternal.WatchConnectionManager - Exec Failure 
java.io.InterruptedIOException: executor rejected
	at okhttp3.RealCall$AsyncCall.executeOn(RealCall.java:186)
	at okhttp3.Dispatcher.promoteAndExecute(Dispatcher.java:186)
	at okhttp3.Dispatcher.enqueue(Dispatcher.java:137)
	at okhttp3.RealCall.enqueue(RealCall.java:127)
	at okhttp3.internal.ws.RealWebSocket.connect(RealWebSocket.java:193)
	at okhttp3.OkHttpClient.newWebSocket(OkHttpClient.java:435)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager.runWatch(WatchConnectionManager.java:156)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager.access$1200(WatchConnectionManager.java:50)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$2$1.execute(WatchConnectionManager.java:320)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$NamedRunnable.run(WatchConnectionManager.java:409)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.concurrent.RejectedExecutionException: Task okhttp3.RealCall$AsyncCall@75a25c7f rejected from java.util.concurrent.ThreadPoolExecutor@4ebe4dd9[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 2]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369)
	at okhttp3.RealCall$AsyncCall.executeOn(RealCall.java:183)
	... 16 more
```

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/Contexts.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/Contexts.kt
@@ -20,13 +20,13 @@ import org.jboss.tools.intellij.kubernetes.model.context.IContext
 interface IContexts {
 	val current: IActiveContext<out HasMetadata, out KubernetesClient>?
 	val all: List<IContext>
-	fun setCurrent(context: IContext)
-	fun clear()
+	fun setCurrent(context: IContext): Boolean
+	fun clear(): Boolean
 }
 
 open class Contexts(
-		private val observable: IModelChangeObservable = ModelChangeObservable(),
-		private val factory: (IModelChangeObservable, NamedContext) -> IActiveContext<out HasMetadata, out KubernetesClient>
+	private val modelObservable: IModelChangeObservable = ModelChangeObservable(),
+	private val factory: (IModelChangeObservable, NamedContext) -> IActiveContext<out HasMetadata, out KubernetesClient>
 ) : IContexts {
 
 	override var current: IActiveContext<out HasMetadata, out KubernetesClient>? = null
@@ -60,61 +60,63 @@ open class Contexts(
 		KubeConfig(::refresh)
 	}
 
-	override fun setCurrent(context: IContext) {
+	override fun setCurrent(context: IContext): Boolean {
 		if (context == current) {
-			return
+			return false
 		}
-		var newContext: IActiveContext<out HasMetadata, out KubernetesClient>?
 		synchronized(this) {
+			if (current != null) {
+				replaceActiveContext(Context(current!!.context), all)
+			}
 			closeCurrent()
-			newContext = create(context.context)
+			val newContext = create(context.context)
 			current = newContext
-			replaceInAllContexts(newContext!!)
+			replaceContext(newContext!!, all)
 		}
-		if (newContext != null) {
-			observable.fireModified(newContext!!)
-		}
+		return true
 	}
 
-	override fun clear() {
-		clear(true)
-	}
-
-	private fun clear(notify: Boolean) {
+	override fun clear(): Boolean {
 		synchronized(this) {
 			all.clear()
-			closeCurrent(notify)
+			return closeCurrent()
 		}
 	}
 
 	protected open fun refresh() {
-		clear(false)
-		observable.fireModified(this) // invalidates root bcs there's no tree node for this
-	}
-
-	private fun closeCurrent() {
-		closeCurrent(true)
-	}
-
-	private fun closeCurrent(notify: Boolean) {
-		val current = this.current
-		current?.close() ?: return
-		this.current = null
-		if (notify) {
-			observable.fireModified(current)
+		if (clear()) {
+			modelObservable.fireModified(this) // invalidates root bcs there's no tree node for this
 		}
+	}
+
+	private fun closeCurrent(): Boolean {
+		val current = this.current
+		current?.close() ?: return false
+		this.current = null
+		return true
 	}
 
 	private fun create(namedContext: NamedContext): IActiveContext<out HasMetadata, out KubernetesClient> {
-		return factory(observable, namedContext)
+		return factory(modelObservable, namedContext)
 	}
 
-	private fun replaceInAllContexts(activeContext: IActiveContext<*, *>) {
-		val contextInAll = all.find { it.context == activeContext.context } ?: return
-		val indexOf = all.indexOf(contextInAll)
-		if (indexOf < 0) {
-			return
-		}
+	private fun replaceContext(activeContext: IActiveContext<*, *>, all: MutableList<IContext>) {
+		val indexOf = indexOf(activeContext.context, all) ?: return
 		all[indexOf] = activeContext
 	}
+
+	private fun replaceActiveContext(context: IContext, all: MutableList<IContext>) {
+		val indexOf = indexOf(context.context, all) ?: return
+		all[indexOf] = context
+	}
+
+	private fun indexOf(context: NamedContext, all: MutableList<IContext>): Int? {
+		val contextInAll = all.find { it.context == context } ?: return null
+		val indexOf = all.indexOf(contextInAll)
+		if (indexOf < 0) {
+			return null
+		}
+		return indexOf
+	}
+
 }

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceModel.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceModel.kt
@@ -43,7 +43,9 @@ open class ResourceModel(
 ) : IResourceModel {
 
     override fun setCurrentContext(context: IContext) {
-        contexts.setCurrent(context)
+        if (contexts.setCurrent(context)) {
+            observable.fireModified(this)
+        }
     }
 
     override fun getCurrentContext(): IActiveContext<out HasMetadata, out KubernetesClient>? {
@@ -83,7 +85,7 @@ open class ResourceModel(
         return ListableCustomResources(definition,this)
     }
 
-    fun <R: HasMetadata> getResources(kind: ResourceKind<R>, namespaced: ResourcesIn, filter: Predicate<R>? = null): Collection<R> {
+    fun <R: HasMetadata> getAllResources(kind: ResourceKind<R>, namespaced: ResourcesIn, filter: Predicate<R>? = null): Collection<R> {
         try {
             val resources: Collection<R> = contexts.current?.getAllResources(kind, namespaced) ?: return emptyList()
             return if (filter == null) {
@@ -99,7 +101,7 @@ open class ResourceModel(
         }
     }
 
-    fun getResources(definition: CustomResourceDefinition): Collection<HasMetadata> {
+    fun getAllResources(definition: CustomResourceDefinition): Collection<HasMetadata> {
         try {
             return contexts.current?.getResources(definition) ?: emptyList()
         } catch(e: IllegalArgumentException) {
@@ -122,8 +124,9 @@ open class ResourceModel(
 
     private fun invalidate() {
         logger<ResourceModel>().debug("Invalidating all contexts.")
-        contexts.clear()
-        observable.fireModified(this)
+        if (contexts.clear()) {
+            observable.fireModified(this)
+        }
     }
 
     private fun invalidate(context: IActiveContext<out HasMetadata, out KubernetesClient>) {

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceModelQuery.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceModelQuery.kt
@@ -36,7 +36,7 @@ class ListableResources<R : HasMetadata>(
 		private val model: ResourceModel) : IListable<R> {
 
 	override fun list(): Collection<R> {
-		return model.getResources(resourceKind, resourceIn, filter)
+		return model.getAllResources(resourceKind, resourceIn, filter)
 	}
 }
 
@@ -45,7 +45,7 @@ class ListableCustomResources(
 		private val model: ResourceModel) : IListable<HasMetadata> {
 
 	override fun list(): Collection<HasMetadata> {
-		return model.getResources(definition)
+		return model.getAllResources(definition)
 	}
 }
 
@@ -60,7 +60,7 @@ class FilterableResources<R : HasMetadata>(
 		private val model: ResourceModel) : IFilterable<R> {
 
 	override fun list(): Collection<R> {
-		return model.getResources(resourceKind, resourceIn)
+		return model.getAllResources(resourceKind, resourceIn)
 	}
 
 	override fun filtered(filter: Predicate<R>): IListable<R> {

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceWatch.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceWatch.kt
@@ -79,7 +79,7 @@ open class ResourceWatch(
         }
     }
 
-    fun clear() {
+    fun close() {
         closeAll(watches.entries.toList())
     }
 

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/context/ActiveContext.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/context/ActiveContext.kt
@@ -364,7 +364,8 @@ abstract class ActiveContext<N : HasMetadata, C : KubernetesClient>(
     }
 
     override fun close() {
-        logger<ActiveContext<*, *>>().debug("Closing context.")
+        logger<ActiveContext<*, *>>().debug("Closing context ${context.name}.")
+        watch.close()
         client.close()
     }
 

--- a/src/test/kotlin/org/jboss/tools/intellij/kubernetes/model/ContextsTest.kt
+++ b/src/test/kotlin/org/jboss/tools/intellij/kubernetes/model/ContextsTest.kt
@@ -10,37 +10,26 @@
  ******************************************************************************/
 package org.jboss.tools.intellij.kubernetes.model
 
-import com.nhaarman.mockitokotlin2.any
-import com.nhaarman.mockitokotlin2.anyOrNull
-import com.nhaarman.mockitokotlin2.clearInvocations
-import com.nhaarman.mockitokotlin2.doReturn
-import com.nhaarman.mockitokotlin2.eq
-import com.nhaarman.mockitokotlin2.mock
-import com.nhaarman.mockitokotlin2.never
-import com.nhaarman.mockitokotlin2.spy
-import com.nhaarman.mockitokotlin2.times
-import com.nhaarman.mockitokotlin2.verify
-import com.nhaarman.mockitokotlin2.whenever
+import com.nhaarman.mockitokotlin2.*
 import io.fabric8.kubernetes.api.model.HasMetadata
 import io.fabric8.kubernetes.api.model.NamedContext
 import io.fabric8.kubernetes.api.model.Namespace
 import io.fabric8.kubernetes.client.KubernetesClient
 import io.fabric8.kubernetes.client.NamespacedKubernetesClient
-import org.assertj.core.api.Assertions
+import org.assertj.core.api.Assertions.assertThat
 import org.jboss.tools.intellij.kubernetes.model.context.IActiveContext
 import org.jboss.tools.intellij.kubernetes.model.context.IContext
 import org.jboss.tools.intellij.kubernetes.model.mocks.ClientMocks
 import org.jboss.tools.intellij.kubernetes.model.mocks.Mocks
+import org.jboss.tools.intellij.kubernetes.model.mocks.Mocks.activeContext
+import org.jboss.tools.intellij.kubernetes.model.mocks.Mocks.context
 import org.junit.Test
 
 class ContextsTest {
 
 	private val client: NamespacedKubernetesClient = mock()
-	private val observable: IModelChangeObservable = mock()
+	private val modelChange: IModelChangeObservable = mock()
 	private val namespace: Namespace = ClientMocks.resource("papa smurf")
-	private val currentContext: IActiveContext<HasMetadata, KubernetesClient> = Mocks.activeContext(client, namespace)
-	private val contextFactory: (IModelChangeObservable, NamedContext?) -> IActiveContext<HasMetadata, KubernetesClient> =
-			Mocks.contextFactory(currentContext)
 	private val namedContext1 =
 			ClientMocks.namedContext("ctx1", "namespace1", "cluster1", "user1")
 	private val namedContext2 =
@@ -48,7 +37,10 @@ class ContextsTest {
 	private val namedContext3 =
 			ClientMocks.namedContext("ctx3", "namespace3", "cluster3", "user3")
 	private val config = createKubeConfig(namedContext2, listOf(namedContext1, namedContext2, namedContext3))
-	private val contexts = spy(TestableContext(observable, contextFactory, config))
+	private val currentContext: IActiveContext<HasMetadata, KubernetesClient> = activeContext(client, namespace, namedContext2)
+	private val contextFactory: (IModelChangeObservable, NamedContext?) -> IActiveContext<HasMetadata, KubernetesClient> =
+		Mocks.contextFactory(currentContext)
+	private val contexts = spy(TestableContext(modelChange, contextFactory, config))
 
 	@Test
 	fun `#clear() should close existing context`() {
@@ -57,26 +49,6 @@ class ContextsTest {
 		contexts.clear()
 		// then
 		verify(currentContext).close()
-	}
-
-	@Test
-	fun `#clear() should notify change of current context`() {
-		// given
-		// when
-		contexts.clear()
-		// then
-		verify(observable).fireModified(currentContext)
-	}
-
-	@Test
-	fun `#clear() should NOT notify change of current context if context is null`() {
-		// given
-		val config = createKubeConfig(null, listOf(namedContext1, namedContext2, namedContext3))
-		val contexts = spy(TestableContext(observable, contextFactory, config))
-		// when
-		contexts.clear()
-		// then
-		verify(observable, never()).fireModified(any())
 	}
 
 	@Test
@@ -112,7 +84,7 @@ class ContextsTest {
 		// when
 		contexts.refresh()
 		// then
-		verify(observable).fireModified(contexts)
+		verify(modelChange).fireModified(contexts)
 	}
 
 	@Test
@@ -166,12 +138,23 @@ class ContextsTest {
     }
 
 	@Test
-	fun `#setCurrentContext(context) should create new active context for given context`() {
+	fun `#setCurrentContext(context) should NOT create new active context if existing context is set`() {
 		// given
-		contexts.current
-		clearInvocations(contextFactory)
+		contexts.current // create current context
+		clearInvocations(contextFactory) // clear invocation so that it's not counted
 		// when
-		contexts.setCurrent(mock())
+		contexts.setCurrent(currentContext)
+		// then
+		verify(contextFactory, never()).invoke(any(), anyOrNull())
+	}
+
+	@Test
+	fun `#setCurrentContext(context) should create new active context if new context is set`() {
+		// given
+		contexts.current // create current context
+		clearInvocations(contextFactory) // clear invocation so that it's not counted
+		// when
+		contexts.setCurrent(context(namedContext3))
 		// then
 		verify(contextFactory).invoke(any(), anyOrNull())
 	}
@@ -179,18 +162,81 @@ class ContextsTest {
 	@Test
 	fun `#setCurrentContext(context) should replace context in #allContexts`() {
 		// given
-		val newCurrentContext = Mocks.activeContext(client, namespace, namedContext3)
+		val newCurrentContext = activeContext(client, namespace, namedContext3)
 		contexts.all // create all contexts
 		val currentContext = contexts.current
-		Assertions.assertThat(currentContext).isNotEqualTo(newCurrentContext)
-		Assertions.assertThat(contexts.all).contains(currentContext)
-		Assertions.assertThat(contexts.all).doesNotContain(newCurrentContext)
+		assertThat(currentContext).isNotEqualTo(newCurrentContext)
+		assertThat(contexts.all).contains(currentContext)
+		assertThat(contexts.all).doesNotContain(newCurrentContext)
 		doReturn(newCurrentContext)
 				.whenever(contextFactory).invoke(any(), anyOrNull()) // returned on 2nd call
 		// when
 		contexts.setCurrent(newCurrentContext)
 		// then
-		Assertions.assertThat(contexts.all).contains(newCurrentContext)
+		assertThat(contexts.all).contains(newCurrentContext)
+	}
+
+	@Test
+	fun `#setCurrentContext(context) should set new current context in #allContexts`() {
+		// given
+		val newCurrentContext = activeContext(client, namespace, namedContext3)
+		contexts.all // create all contexts
+		val currentContext = contexts.current
+		assertThat(currentContext).isNotEqualTo(newCurrentContext)
+		assertThat(contexts.all).contains(currentContext)
+		assertThat(contexts.all).doesNotContain(newCurrentContext)
+		doReturn(newCurrentContext)
+			.whenever(contextFactory).invoke(any(), anyOrNull()) // returned on 2nd call
+		// when
+		contexts.setCurrent(newCurrentContext)
+		// then
+		assertThat(contexts.all).contains(newCurrentContext)
+	}
+
+	@Test
+	fun `#setCurrentContext(context) should remove current context in #allContexts`() {
+		// given
+		val newCurrentContext = activeContext(client, namespace, namedContext3)
+		contexts.all // create all contexts
+		val currentContext = contexts.current
+		assertThat(currentContext).isNotEqualTo(newCurrentContext)
+		assertThat(contexts.all).contains(currentContext)
+		assertThat(contexts.all).doesNotContain(newCurrentContext)
+		doReturn(newCurrentContext)
+			.whenever(contextFactory).invoke(any(), anyOrNull()) // returned on 2nd call
+		// when
+		contexts.setCurrent(newCurrentContext)
+		// then
+		assertThat(contexts.all).doesNotContain(currentContext)
+	}
+
+	@Test
+	fun `#setCurrentContext(context) should close current context`() {
+		// given
+		val newCurrentContext = activeContext(client, namespace, namedContext3)
+		contexts.all // create all contexts
+		val currentContext = contexts.current!!
+		doReturn(newCurrentContext)
+			.whenever(contextFactory).invoke(any(), anyOrNull()) // returned on 2nd call
+		// when
+		contexts.setCurrent(newCurrentContext)
+		// then
+		verify(currentContext).close()
+	}
+
+	@Test
+	fun `#setCurrentContext(context) should return 'true' if new context was set`() {
+		// given
+		val newCurrentContext = activeContext(client, namespace, namedContext3)
+		contexts.all // create all contexts
+		val currentContext = contexts.current!!
+		doReturn(newCurrentContext)
+			.whenever(contextFactory).invoke(any(), anyOrNull()) // returned on 2nd call
+		// when
+		val currentContextIsSet = contexts.setCurrent(newCurrentContext)
+		// then
+		assertThat(contexts.current).isEqualTo(newCurrentContext)
+		assertThat(currentContextIsSet).isTrue()
 	}
 
     private fun createKubeConfig(currentContext: NamedContext?, allContexts: List<NamedContext>): KubeConfig {
@@ -203,10 +249,10 @@ class ContextsTest {
         }
     }
 
-	private class TestableContext(observable: IModelChangeObservable,
+	private class TestableContext(modelChange: IModelChangeObservable,
 								  factory: (IModelChangeObservable, NamedContext) -> IActiveContext<out HasMetadata, out KubernetesClient>,
 								  override val config: KubeConfig
-	): Contexts(observable, factory) {
+	): Contexts(modelChange, factory) {
 
 		override val all: MutableList<IContext> = spy(super.all)
 

--- a/src/test/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceModelTest.kt
+++ b/src/test/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceModelTest.kt
@@ -48,7 +48,7 @@ class ResourceModelTest {
     private val client: NamespacedKubernetesClient = mock()
     private val modelChange: IModelChangeObservable = mock()
     private val namespace: Namespace = resource("papa smurf")
-    private val activeContext: IActiveContext<HasMetadata, KubernetesClient> = activeContext(client, namespace)
+    private val activeContext: IActiveContext<HasMetadata, KubernetesClient> = activeContext(client, namespace, mock())
     private val contextFactory: (IModelChangeObservable, NamedContext?) -> IActiveContext<HasMetadata, KubernetesClient> =
             contextFactory(activeContext)
 
@@ -70,7 +70,7 @@ class ResourceModelTest {
     fun `#getResources(kind, CURRENT_NAMESPACE) should call context#getResources(kind, CURRENT_NAMESPACE)`() {
         // given
         // when
-        model.getResources(NamespacedPodsProvider.KIND, ResourcesIn.CURRENT_NAMESPACE)
+        model.getAllResources(NamespacedPodsProvider.KIND, ResourcesIn.CURRENT_NAMESPACE)
         // then
         verify(activeContext).getAllResources(NamespacedPodsProvider.KIND, ResourcesIn.CURRENT_NAMESPACE)
     }
@@ -79,7 +79,7 @@ class ResourceModelTest {
     fun `#getResources(kind, NO_NAMESPACE) should call context#getResources(kind, NO_NAMESPACE)`() {
         // given
         // when
-        model.getResources(NamespacedPodsProvider.KIND, ResourcesIn.NO_NAMESPACE)
+        model.getAllResources(NamespacedPodsProvider.KIND, ResourcesIn.NO_NAMESPACE)
         // then
         verify(activeContext).getAllResources(NamespacedPodsProvider.KIND, ResourcesIn.NO_NAMESPACE)
     }
@@ -94,7 +94,7 @@ class ResourceModelTest {
                         POD2,
                         POD3))
         // when
-        model.getResources(NamespacedPodsProvider.KIND, ResourcesIn.NO_NAMESPACE, filter)
+        model.getAllResources(NamespacedPodsProvider.KIND, ResourcesIn.NO_NAMESPACE, filter)
         // then
         verify(filter, times(3)).test(any())
     }
@@ -104,7 +104,7 @@ class ResourceModelTest {
         // given
         val definition = mock<CustomResourceDefinition>()
         // when
-        model.getResources(definition)
+        model.getAllResources(definition)
         // then
         verify(activeContext).getResources(definition)
     }
@@ -137,12 +137,25 @@ class ResourceModelTest {
     }
 
     @Test
-    fun `#invalidate(model) should fire model moidified`() {
+    fun `#invalidate(model) should fire 1x model modified if contexts was cleared`() {
         // given
+        whenever(contexts.clear())
+            .doReturn(true)
         // when
         model.invalidate(model)
         // then
-        verify(modelChange).fireModified(model)
+        verify(modelChange, times(1)).fireModified(model)
+    }
+
+    @Test
+    fun `#invalidate(model) should NOT fire model modified if contexts were cleared`() {
+        // given
+        whenever(contexts.clear())
+            .doReturn(false)
+        // when
+        model.invalidate(model)
+        // then
+        verify(modelChange, never()).fireModified(model)
     }
 
     @Test
@@ -216,6 +229,30 @@ class ResourceModelTest {
         model.setCurrentContext(currentContext!!)
         // then
         verify(contextFactory, never()).invoke(any(), anyOrNull())
+    }
+
+    @Test
+    fun `#setCurrentContext(context) should fire if 'contexts' was successfully set new context`() {
+        // given
+        whenever(contexts.setCurrent(any()))
+            .doReturn(true)
+        // when
+        model.setCurrentContext(mock())
+        // then
+        verify(modelChange, times(1))
+            .fireModified(model)
+    }
+
+    @Test
+    fun `#setCurrentContext(context) should NOT fire if 'contexts' failed to set new context`() {
+        // given
+        whenever(contexts.setCurrent(any()))
+            .doReturn(false)
+        // when
+        model.setCurrentContext(mock())
+        // then
+        verify(modelChange, never())
+            .fireModified(model)
     }
 
     private fun createContexts(currentContext: IActiveContext<*,*>?, allContexts: List<IContext>): IContexts {

--- a/src/test/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceWatchTest.kt
+++ b/src/test/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceWatchTest.kt
@@ -204,10 +204,10 @@ class ResourceWatchTest {
     }
 
     @Test
-    fun `#clear() should close existing watches`() {
+    fun `#close() should close existing watches`() {
         // given
         // when starting 2nd time
-        resourceWatch.clear()
+        resourceWatch.close()
         // then
         assertThat(watchable1.watch.isClosed())
         assertThat(watchable2.watch.isClosed())

--- a/src/test/kotlin/org/jboss/tools/intellij/kubernetes/model/mocks/Mocks.kt
+++ b/src/test/kotlin/org/jboss/tools/intellij/kubernetes/model/mocks/Mocks.kt
@@ -49,7 +49,7 @@ object Mocks {
         }
     }
 
-    fun activeContext(client: NamespacedKubernetesClient, currentNamespace: Namespace, context: NamedContext? = null)
+    fun activeContext(client: NamespacedKubernetesClient, currentNamespace: Namespace, context: NamedContext)
             : IActiveContext<HasMetadata, KubernetesClient> {
         return mock {
             doReturn(client)

